osbuild/__init__.py

```diff
@@ -69,13 +70,156 @@ def __exit__(self, exc_type, exc_value, exc_tb):
         self.root = None
 
 
+def upsum(m, dir_fd):
+    """Compute a content hash of a filesystem tree
+
+    The hash is stable between runs, and guarantees that two filesystem
+    trees with the same hash, are functionally equivalent from the OS
+    point of view.
+
+    The file, symlink and directory names and contents are recursively
+    hashed, together with security-relevant metadata."""
+
+    with os.scandir(dir_fd) as it:
+        for dirent in sorted(it, key=(lambda d: d.name)):
+            stat_result = dirent.stat(follow_symlinks=False)
+            metadata = {}
+            metadata["name"] = os.fsdecode(dirent.name)
+            metadata["mode"] = stat_result.st_mode
+            metadata["uid"] = stat_result.st_uid
+            metadata["gid"] = stat_result.st_gid
+            # include the size of symlink target/file-contents so we don't have to delimit it
+            metadata["size"] = stat_result.st_size
+            # getxattr cannot operate on a dir_fd, so do a trick and rely on the entries in /proc
+            stable_file_path = os.path.join(f"/proc/self/fd/{dir_fd}", dirent.name)
```

Comment:
Why are you using `os.path.join` for one part of the path and an `f`-string for another part?

Reply:
No particular good reason, but here it is: I'm happy to create a string representation for a path from strings or ints, but in case `dirent.name` is a bytes object (which I admit it may or may not be here), I didn't want to get into the business of decoding it (as it is not necessarily valid UTF-8, and everything is rather involved, right?), but I figured that `os.path.join()` would do this "right", so that I get a path out that I am guaranteed to be pointing at the right file. Most likely it would be fine either way :)

Reply:
It does the right thing, but you're now passing a path as a string to `os.getxattr`, not as a `bytes`. I think that doesn't matter, but you used `fsencode` everywhere else.

Reply:
Let's address this separately.